### PR TITLE
feat: incremental static regeneration

### PR DESCRIPTION
**Pull request by andykenward**
See if we can use incremental static regeneration for next.js pages too improve page load.

We may be able to use this for doc pages too.

https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration

Relates to #33

### CODE REVIEW

**Review by patrick91 — APPROVED** [review]

**Comment by patrick91** [comment]
This is neat, the CoC and Acknowledgements page load super fast now! for the docs I guess it is a bit tricky, since we allow to preview docs, like here: https://github.com/strawberry-graphql/strawberry/pull/687#issuecomment-767645898

maybe we can have a dynamic route just for that use case? 🤔

**Comment by andykenward** [comment]
Perhaps we can use [next.js preview mode](
https://nextjs.org/docs/advanced-features/preview-mode) feature for PR?

**Comment by andykenward** [comment]
Going to spend some time on the [docs/[[...slug]].ts](https://github.com/strawberry-graphql/strawberry.rocks/blob/499f02ae00bf2ec9771d4c446d022502ecca3590/pages/docs/%5B%5B...slug%5D%5D.tsx#L112-L162) pages to also use incremental static regeneration. 
Need to use [getstaticpaths](https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation) to generate all the docs page paths to render.

**Comment by patrick91** [comment]
> Perhaps we can use [next.js preview mode](https://nextjs.org/docs/advanced-features/preview-mode) feature for PR?

that could be an option, but I never tried it, not sure how it works :)

**Comment by andykenward** [comment]
@patrick91 Should all the GitHub API queries for doc files be against the latest release tag and not master?
I am planning to allow the website to display different release version docs later on e.g `/docs/tag/1.0.0/general/schema-basics`. But I think the default for `/docs/general/schema-basics` should show the latest release of strawberry.

**Comment by patrick91** [comment]
> @patrick91 Should all the GitHub API queries for doc files be against the latest release tag and not master?

Good point! We can do latest tag :)


> I am planning to allow the website to display different release version docs later on e.g `/docs/tag/1.0.0/general/schema-basics`. But I think the default for `/docs/general/schema-basics` should show the latest release of strawberry.

Sounds perfect 🙌🏼

**Comment by andykenward** [comment]
TODO: 
 - pull request support refactor. Going to add a another nextjs page of `/docs/pr/[[...slug]].ts` as the caching and static generation will be different for theses. Having it all in one page is a  bit much.
 - tag/release support for older versions of the docs. This will also have its own nextjs page of `/docs/tag/[[...slug]].ts` as this will be static generated at build and shouldn't need revalidation.

**Comment by andykenward** [comment]
The docs should be a lot faster. There is a flash of white when you first hit the docs index page. I think it is a nextjs issue as it doesn't generate the static page for `[[...slug]].html` with the content of the `docs/index.md`. It doesnt execute the getStaticProps function when `params:{}`. The docs say to add `[]` for slug to `getStaticPaths`. Need to look into it more.

**Comment by andykenward** [comment]
Refactored the pull request doc preview.

Visiting `/docs/pr/${pull_request_number}` that has a github label of `ok-to-preview` will render on the site. If the pull request is not labelled then 404s.

A pull request doc pages are rendered on the backend on demand and not at build time.

**Comment by andykenward** [comment]
Visiting `docs/tag/${tag}` will render on demand the docs for that version. 

But this will use the current design of the site. So older doc markdown files might not render as expect? 

If we decided to change the site structure or design we might have to keep legacy code to support older versions. Or add logic to limit the tag versions the site will support and direct users to the github repo docs folder for that version instead.

**Comment by patrick91** [comment]
> Visiting `docs/tag/${tag}` will render on demand the docs for that version. There is no next.js revalidation of these pages, as they should never need to be rerender after the first request. I expect if the site gets redeployed then they will?

Sounds good :)

> But this will use the current design of the site. So older doc markdown files might not render as expect?

This should be fine. I think we might ignore versions less than v1 in the dropdown list. We have a lot of releases at the moment. I actually wonder if we should only do docs for x.y, instead of x.y.z, what do you think? /cc @BryceBeagle 

> If we decided to change the site structure or design we might have to keep legacy code to support older versions. Or add logic to limit the tag versions the site will support and direct users to the github repo docs folder for that version instead.

That's an option too, but I don't see us using a lot of fancy things in markdown, so I think should be fine :)

File: pages/docs/pr/[[...slug]].tsx
Comment (patrick91) [inline pages/docs/pr/[[...slug]].tsx:39]: don't think we'd need ^^

File: pages/docs/pr/[[...slug]].tsx
Comment (patrick91) [inline pages/docs/pr/[[...slug]].tsx:50]: nice!

File: lib/api.ts
Comment (patrick91) [inline lib/api.ts:106]: yes, this could be a good idea, for now the links are only visible when adding the label, so we can keep it like that :)

File: lib/api.ts
Comment (patrick91) [inline lib/api.ts:117]: shall we set sentry for all these errors?

File: lib/api.ts
Comment (patrick91) [inline lib/api.ts:37]: 💯

**Review by patrick91 — APPROVED** [review]
This is amazing work! Thank you so much ^^

File: helpers/image-paths.ts
Comment (andykenward) [reply helpers/image-paths.ts:6]: Going to refactor this to use an object for the functions parameters

Comment (andykenward) [reply helpers/image-paths.ts]: This is a great feature in vscode! Even implemented the change to the usages.
CleanShot 2021-02-18 at 11 03 26@2x

**Comment by BryceBeagle** [comment]
> I actually wonder if we should only do docs for x.y, instead of x.y.z, what do you think? /cc @BryceBeagle

Historically we've had documentation-only changes in releases that only bump the patch version, or no bump at all. This is fine if we're okay with just waiting for the next x.y release for the documentation improvements to make their way to the website.

File: lib/api.ts
Comment (patrick91) [reply lib/api.ts:117]: we can also implement it later 😊

**Comment by andykenward** [comment]
@patrick91 Is there anything more we need to do before this pull request can be merged?

I've made a few more changes since the last review.

- converted all but the contributors GitHub API calls from REST to GraphQL now.
- generate TypeScript typings for GraphQL Queries using GraphQL Code Generator.
- enabled TypeScript strict mode.
- `npm run build` cli now has a `prebuild` that generates the GraphQL typings and runs a type check.
- updated README.md.

File: README.md
Comment (patrick91) [inline README.md:14]: Nice!

**Review by patrick91 — APPROVED** [review]
Looks great! Will merge when I get home later today ☺️

**Pull request merged**
